File: bench/gcscan/gcscan.go
// GC-scan-cost, Go: the 2M live strings are pointers the GC re-scans every cycle.
package main
import ("fmt"; "os"; "runtime"; "strconv")
func main() {
    const m = 2000000
    live := make([]string, m)
    for i := 0; i < m; i++ { live[i] = "x" + strconv.Itoa(i%1000) }
    acc := 0
    for r := 0; r < 300; r++ {
        xs := make([]int, 0)
        for j := 0; j < 50000; j++ { xs = append(xs, (r+j)%997) }
        s := 0; for _, v := range xs { s += v }
        acc = (acc + s) % 1000000007
    }
    lacc := 0; for i := 0; i < m; i++ { lacc += len(live[i]) }
    var ms runtime.MemStats; runtime.ReadMemStats(&ms)
    fmt.Fprintf(os.Stderr, "gc=%d pause_us=%d\n", ms.NumGC, ms.PauseTotalNs/1000)
    fmt.Println(acc, lacc)
}

File: bench/gcscan/gcscan.c
/* GC-scan-cost, C: 2M strings held in an array (no GC, never scanned) + churn. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int main(void) {
    long m = 2000000;
    char **live = malloc((size_t)m * sizeof(char *));
    for (long i = 0; i < m; i++) { char b[16]; snprintf(b, sizeof b, "x%ld", i % 1000); live[i] = strdup(b); }
    long acc = 0;
    for (long r = 0; r < 300; r++) {
        long *xs = malloc((size_t)50000 * sizeof(long));
        for (long j = 0; j < 50000; j++) xs[j] = (r + j) % 997;
        long s = 0; for (long t = 0; t < 50000; t++) s += xs[t];
        acc = (acc + s) % 1000000007L;
        free(xs);
    }
    long lacc = 0; for (long i = 0; i < m; i++) lacc += (long)strlen(live[i]);
    printf("%ld %ld\n", acc, lacc);
    return 0;
}
